### PR TITLE
ENH: Directly Implement Frobenius Norms for Vectors

### PR DESCRIPTION
**Pull request by TNonet**
Default parameters for norms implement the Frobenius Norm. See #15533

When using ord='fro' should work for both vectors and arrays, as defined.

### CODE REVIEW

File: numpy/linalg/linalg.py
Comment (eric-wieser) [inline numpy/linalg/linalg.py]: It's tempting to handle this on line 2538, as:
```
# frobenius is equivalent to the 2-norm, which is the default
if ord is None or ord in ('f', 'fro'):
    ord = 2
```
which makes this line just `ord == 2`

Comment (TNonet) [reply numpy/linalg/linalg.py]: I think that would work as well. I am unsure which would be the superior option. I am rather new to submitting pull requests so I might defer to your judgment or more feedback.

Comment (rossbar) [reply numpy/linalg/linalg.py]: I agree with @eric-wieser - organizing it in this way makes explicit the relationship with the default behavior

File: numpy/linalg/tests/test_linalg.py
Comment (rossbar) [inline numpy/linalg/tests/test_linalg.py]: Hmmm - the inclusion of these tests implies that the failure of `ord='fro'` was intentional for 1D vectors. I'm not sure why this would be the case, as the default behavior (i.e. `ord=None`) is equivalent to `ord='fro'`. I'm no expert in vector math, but [that wolfram article](http://mathworld.wolfram.com/FrobeniusNorm.html) originally cited by @TNonet makes it seem that the equivalence between "Frobenius norm" and the L2 vector norm is colloquially accepted but perhaps not strictly true. Maybe @seberg remembers if there was ever conversation along these lines related to #14719 etc.?

Either way, my vote would be for keeping these proposed changes, as they reflect the behavior of the code (i.e. default is equivalent to Frobenius even for the 1D case).

File: numpy/linalg/tests/test_regression.py
Comment (rossbar) [inline numpy/linalg/tests/test_regression.py]: This is strange: #786 doesn't seem to be a meaningful referral. I checked back in the PRs/Issues on github, and issue #786 has nothing to do with norms. It seems a lot of issues/PRs from this era were created with a different system, so this does not seem to be particularly helpful in resolving my above comment.

Either way, I agree this test should be removed if we go forward with the current approach

**Comment by seberg** [comment]
Looking at it again, is "Frobenius" norm actually used for vectors in mathematics, or are we generalizing it to vectors even though the name is normally limited to matrices?

File: numpy/linalg/linalg.py
Comment (rossbar) [reply numpy/linalg/linalg.py]: I agree with @eric-wieser - his proposed approach is more explicit in demonstrating the relationship between the default behavior (i.e. `ord=None`) and `ord='fro'`.

Comment (rossbar) [reply numpy/linalg/linalg.py]: While we're at it - line 2553 is difficult to grok and seems to be the exception raising mechanism for `ord='fro'` for 1D vectors. It should probably be cleaned up to be consistent with the other changes in this PR.

**Review by rossbar — CHANGES_REQUESTED** [review]
Looks like a great start @TNonet , thanks for getting this started!

In addition to addressing the inline comments - the docstring should be updated as well. For example, the table that describes how the various norms are calculated for matrices/vectors should have the "norm for vectors" column updated for the 'fro' keyword. A pass over the whole docstring wouldn't hurt either.

Finally, it might be possible that the current behavior (that we are changing) is depended on downstream, e.g. in scipy.linalg.norm. Perhaps @WarrenWeckesser could weigh in on whether these changes make sense or cause consistency issues?

**Comment by eric-wieser** [comment]
> are we generalizing it to vectors even though the name is normally limited to matrices?

This is a good question - I think I'd like to see some evidence that this terminology makes sense before merging it.

File: numpy/linalg/tests/test_regression.py
Comment (eric-wieser) [reply numpy/linalg/tests/test_regression.py]: This is why I've been leaning towards `gh-NNN` style referencing, so that if this happens again we're less confused.

Comment (eric-wieser) [reply numpy/linalg/tests/test_regression.py]: You're looking for https://github.com/numpy/numpy/issues/1384

Comment (rossbar) [reply numpy/linalg/tests/test_regression.py]: Thanks @eric-wieser . The conversation there doesn't appear to have any definitive conclusions, just noting that this issue was discovered way back in 2008

**Comment by seberg** [comment]
Otherwise noting that these are matrix norms in the doc (better) and making the error obvious would be good though.

File: numpy/linalg/tests/test_linalg.py
Comment (TNonet) [reply numpy/linalg/tests/test_linalg.py]: It seems that [MathWorks](https://www.mathworks.com/help/matlab/ref/norm.html), implements the Frobenius Norm for an m by n matrices where m and n are both greater or equal to 2. In addition, they directly implement [VectorNorms](https://www.mathworks.com/help/matlab/ref/vecnorm.html) which doesn't have Frobenius Norms.

However, [NetLib](https://www.netlib.org/lapack/lug/node75.html) states the 2-norm and the Frobenius Norms are equivalent for vectors.

It seems like there is no standard.

Therefore, I think that as long as no downstream code is adversely affected, I agree with @rossbar that the change is reflected in the behavior of the code.

Comment (rossbar) [reply numpy/linalg/tests/test_linalg.py]: OTOH we certainly don't want to promote incorrect/imprecise use of terminology. In the absence of definitive, unambiguous resolution of the terminology, it is probably best to keep the current behavior but to clean up the docstring and the error messages so that the distinction between Frobenius (relevant ONLY for matrices) and L2-norm (default behavior for vectors) explicit.

Comment (seberg) [reply numpy/linalg/tests/test_linalg.py]: I am not quite sure, I tend to: Leaving it as is seems good, but the error message should be clear. Note that there is a nice table in the `Notes` section of the docstring, which is very clear on the implemented meaning.

**Comment by TNonet** [comment]
So the consensus seems to be to close this pull request and resubmit a new one with an updated error message that is more descriptive and potentially add clarity to the docstring.

**Comment by eric-wieser** [comment]
@TNonet: That would be my vote, yes

**Comment by rossbar** [comment]
ping @TNonet are you still working on this PR? It's currently in draft status but if you were planning on closing please feel free to do so.

**Comment by TNonet** [comment]
> ping @TNonet are you still working on this PR? It's currently in draft status but if you were planning on closing please feel free to do so.

Sorry. I am embarrassed that I let this drop off my radar. I will be closing this PR. The conversation has continued on #15740

**Comment by rossbar** [comment]
No worries, thanks for helping to work through it and figure out what the path forward should be!

**Pull request closed without merging**
